### PR TITLE
Update the Cucumber Test to create schema and verify its creation

### PR DESCRIPTION
**Pull request by Wibulization**

### CODE REVIEW

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java [hunk 1]
```diff
@@ -1,25 +1,101 @@
...
+    private JdbcSchemaRepository schema;
+    private HikariDataSource dataSource;
+
     @Given("^a Postgres database without schemas$")
-    public void aSchemaWithATable() {
+    public void aSchemaWithATable() throws Exception {
+        // Create Database
+        CreatePostgres.getInstance();
+        dataSource = dataSource();
+        dropSchemaIfExist();
```
Comment (peterbean410) [inline engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]:
```suggestion
        dropSchemaIfExists();
```

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java [hunk 2]
```diff
@@ -1,25 +1,101 @@
...
 import io.cucumber.java.en.When;
 import io.cucumber.java.en.Then;
 
 public class DatabaseDefinitionSteps {
+    private JdbcSchemaRepository schema;
+    private HikariDataSource dataSource;
+
     @Given("^a Postgres database without schemas$")
-    public void aSchemaWithATable() {
+    public void aSchemaWithATable() throws Exception {
```
Comment (peterbean410) [inline engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]:
```suggestion
    public void aPostgresDatabaseWithoutSchemas() throws Exception {
```

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java [hunk 3]
```diff
@@ -1,25 +1,101 @@
...
+        // Create Database
+        CreatePostgres.getInstance();
+        dataSource = dataSource();
+        dropSchemaIfExist();
     }
 
-
     @And("an implementation of SchemaRepository")
     public void anImplementationOfSchemaRepository() {
+        schema = implementationOfSchema();
```
Comment (peterbean410) [inline engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]: As discussed

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java [hunk 4]
```diff
@@ -1,25 +1,101 @@
...
     }
 
     @When("createBaseTables\\() is executed")
     public void createbasetablesIsExecuted() {
+        schema.createBaseTables();
     }
 
     @Then("schema datorum_schema is created")
     public void schemaIsCreated() {
+        checkSchemaExist();
```
Comment (peterbean410) [inline engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]: Let's put schemaName as a string param here

File: engine/specs-java/io/beandev/datorum/CreatePostgres.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            V1Pod pod = api.readNamespacedPod(podName, namespace, null);
+            if (pod.getStatus() != null && pod.getStatus().getPhase().equals("Running")) {
+                System.out.println("Pod is ready.");
+                break;
+            } else {
+                System.out.println("Pod is not ready, waiting...");
+                TimeUnit.SECONDS.sleep(30);
+            }
+        }
+    }
```
Comment (peterbean410) [inline engine/specs-java/io/beandev/datorum/CreatePostgres.java]: Let's put 1 minute as TIMEOUT

**Review by peterbean410 — CHANGES_REQUESTED** [review]
Let's also update the Makefile to add `test-engine` rule such that it will run the test for `engine` Gradle sub project.

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java (hunk 3 above)
Comment (Wibulization) [reply engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]: image

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java (hunk 4 above)
Comment (Wibulization) [reply engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]: image

File: engine/specs-java/io/beandev/datorum/CreatePostgres.java (hunk 5 above)
Comment (Wibulization) [reply engine/specs-java/io/beandev/datorum/CreatePostgres.java]: image

**Comment by Wibulization** [comment]
> Let's also update the Makefile to add `test-engine` rule such that it will run the test for `engine` Gradle sub project.
image

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java (hunk 4 above)
Comment (Wibulization) [reply engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]: Change datorum_schema to word expression to get the name of schema. Then we get schemaName as a parameter of method checkSchemaExist()

Comment (Wibulization) [reply engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]: Here is the method checkSchemaExist when it has parameter
image

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java (hunk 3 above)
Comment (Wibulization) [reply engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]: We don't need the method implementationOfSchema to create schema repository, we just create it directly

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java (hunk 2 above)
Comment (Wibulization) [reply engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]: Change the name of the method

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java (hunk 1 above)
Comment (Wibulization) [reply engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]: Change the name of the method

Comment (Wibulization) [reply engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]: image
image

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java (hunk 2 above)
Comment (Wibulization) [reply engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]: image

File: engine/specs-java/io/beandev/datorum/CreatePostgres.java (hunk 5 above)
Comment (Wibulization) [reply engine/specs-java/io/beandev/datorum/CreatePostgres.java]: image

File: engine/specs-java/io/beandev/datorum/CreatePostgres.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        checkServiceExistsAndCreateService(api, "default", "postgres-service");
+
+        // Wait for the Service to be ready
+        waitForServiceReady(api, "default", "postgres-service");
+
+        // Check if Pod already exists
+        checkPodExistsAndCreatePod(api, "default", "postgres");
+
+        // Wait for the Pod to be ready
+        waitForPodReady(api, "default", "postgres");
```
Comment (peterbean410) [inline engine/specs-java/io/beandev/datorum/CreatePostgres.java]: Check again and remove all unnecessary comments, try to name the method as clear and descriptive as possible.

File: engine/specs-java/io/beandev/datorum/CreatePostgres.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    private static void checkServiceExistsAndCreateService(CoreV1Api api, String namespace, String serviceName)
+            throws Exception {
+        try {
+            V1Service existingService = api.readNamespacedService(serviceName, namespace, null);
+            System.out.println("Service already exists: " + existingService.getMetadata().getName());
+        } catch (ApiException e) {
+            if (e.getCode() == 404) {
+                // Service does not exist, create it
+                V1Service service = createService();
+                V1Service createdService = api.createNamespacedService(namespace, service, null, null, null, null);
```
Comment (peterbean410) [inline engine/specs-java/io/beandev/datorum/CreatePostgres.java]: Which line actually creates the service ? L170 or L171?

If L170 does not actually create the service, why did you name it `createService()` ?

File: engine/specs-java/io/beandev/datorum/CreatePostgres.java (hunk 6 above)
Comment (Wibulization) [reply engine/specs-java/io/beandev/datorum/CreatePostgres.java]: image

File: engine/specs-java/io/beandev/datorum/CreatePostgres.java (hunk 7 above)
Comment (Wibulization) [reply engine/specs-java/io/beandev/datorum/CreatePostgres.java]: image

Comment (Wibulization) [reply engine/specs-java/io/beandev/datorum/CreatePostgres.java]: image

File: engine/specs-java/io/beandev/datorum/CreatePostgres.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+    // Method to ensure that pod exists
+    private static void ensurePodExists(CoreV1Api api, String namespace, String podName) throws Exception {
+        try {
+            V1Pod existingPod = api.readNamespacedPod(podName, namespace, null);
+            System.out.println("Pod already exists: " + existingPod.getMetadata().getName());
+        } catch (ApiException e) {
+            if (e.getCode() == 404) {
+                // Pod does not exist, create it
+                V1Pod podBody = podBody();
```
Comment (peterbean410) [inline engine/specs-java/io/beandev/datorum/CreatePostgres.java]:
```suggestion
                V1Pod postgresPodDef = createPostgresPodDefinition();
```

Comment (Wibulization) [reply engine/specs-java/io/beandev/datorum/CreatePostgres.java]: image

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java [hunk 9]
```diff
@@ -1,25 +1,96 @@
...
+        CreatePostgres.getInstance();
+        dataSource = dataSource();
+        dropSchemaIfExists();
     }
 
-
     @And("an implementation of SchemaRepository")
     public void anImplementationOfSchemaRepository() {
+        jdbcSchemaRepository = new JdbcSchemaRepository(dataSource);
+        Assertions.assertNotNull(jdbcSchemaRepository, "JdbcSchemaRepository should not null");
```
Comment (peterbean410) [inline engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]:
```suggestion
```
Are you trying it test if Java `new` keyword works?

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java [hunk 10]
```diff
@@ -1,25 +1,96 @@
...
+        } catch (SQLException ex) {
+            ex.printStackTrace();
+        } finally {
+            if (dataSource != null) {
+                dataSource.close();
+            }
+        }
+    }
+
+    private void dropSchemaIfExists() {
```
Comment (peterbean410) [inline engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]:
```suggestion
    private void dropSchemaIfExists(schemaName: String) {
```

File: engine/specs-java/io/beandev/datorum/CreatePostgres.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+                V1Pod createdPod = api.createNamespacedPod(namespace, postgresPodDef, null, null, null, null);
+                System.out.println("PostgreSQL Pod created: " + createdPod.getMetadata().getName());
+            } else {
+                throw e; // Rethrow other API exceptions
+            }
+        }
+    }
+
+    // Method to ensure that service exists
+    private static void ensureServiceExists(CoreV1Api api, String namespace, String serviceName)
```
Comment (peterbean410) [inline engine/specs-java/io/beandev/datorum/CreatePostgres.java]: ensure what service ?

File: engine/src/testUnit/java/io/beandev/datorum/migration/MigratorTest.java [hunk 12]
```diff
@@ -17,4 +17,10 @@ void myTest() {
         out.println("hello");
         out.println("Testing " + repository.findAggregateSnapshot(null));
     }
+
+    @Test
+    void myTest2() {
+        out.println("hello");
+        out.println("Testing " + repository.findAggregateSnapshot(null));
+    }
```
Comment (peterbean410) [inline engine/src/testUnit/java/io/beandev/datorum/migration/MigratorTest.java]: ???????

Comment (Wibulization) [reply engine/src/testUnit/java/io/beandev/datorum/migration/MigratorTest.java]: image
Remove unnecessary test

File: engine/specs-java/io/beandev/datorum/CreatePostgres.java (hunk 11 above)
Comment (Wibulization) [reply engine/specs-java/io/beandev/datorum/CreatePostgres.java]: image

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java (hunk 10 above)
Comment (Wibulization) [reply engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]: image
Method
image

File: engine/specs-java/steps/schema/DatabaseDefinitionSteps.java (hunk 9 above)
Comment (Wibulization) [reply engine/specs-java/steps/schema/DatabaseDefinitionSteps.java]: image

**Review by peterbean410 — APPROVED** [review]

**Pull request merged**